packages/flutter/lib/src/widgets/value_updater.dart

```diff
@@ -0,0 +1,171 @@
+/// pressed. This is done without needing to create a stateful widget.
+///
+/// ** See code in examples/api/lib/widgets/value_updater/value_updater.1.dart **
+/// {@end-tool}
+
+class ValueUpdater<T> extends StatefulWidget {
+  /// Creates a [ValueUpdater] widget that updates the value based on
+  /// [valueNotifier] changes.
+  ///
+  /// The [builder], [initialValue], and [valueNotifier] are
+  /// required.
+  const ValueUpdater({
+    super.key,
+    required this.builder,
+    required this.initialValue,
+    this.valueNotifier,
+  })  : value = null;
+
+  /// Creates a [ValueUpdater] widget that updates the value only once (the
+  /// first time it is built) based on [value].
+  ///
+  /// The [builder], [initialValue], and [value] are required.
+  const ValueUpdater.once({
+    super.key,
+    required this.builder,
+    required this.initialValue,
+    this.value,
+  })  : valueNotifier = null;
+
+  /// A builder function which takes a [BuildContext], and the current [value]
+  /// (or the value of the [valueNotifier]) and returns a widget.
+  ///
+  /// Required.
+  final ValueUpdaterBuilder<T> builder;
+
+  /// The value to pass to the [builder] when the widget is first built.
+  ///
+  /// Required.
+  final T initialValue;
+
+  /// The optional value to pass to the [builder] when the widget is built the
+  /// second and subsequent times.
+  ///
+  /// Ignored if [valueNotifier] is set.
+  ///
+  /// If not specified, then the current value of [valueNotifier] is used. One
+  /// of [value] or [valueNotifier] must be set.
+  ///
+  /// Only one of [valueNotifier] or [value] may be set at a time, but one of
+  /// them must be set.
+  final T? value;
+
+  /// The optional [ValueNotifier] to listen to.
+  ///
+  /// If specified, instead toggling between [initialValue] and [value] only
+  /// once, whenever the [ValueNotifier] notifies of a change, the widget will
+  /// use the notifier's value as the next value to build the [builder] with.
+  ///
+  /// Only one of [valueNotifier] or [value] may be set at a time, but one of
+  /// them must be set.
+  final ValueNotifier<T>? valueNotifier;
+
+  @override
+  State<ValueUpdater<T>> createState() => _ValueUpdaterState<T>();
+}
+
+enum _ValueUpdaterStatus {
+  initial,
+  off,
+  on,
+}
+
+class _ValueUpdaterState<T> extends State<ValueUpdater<T>> {
+  _ValueUpdaterStatus _triggered = _ValueUpdaterStatus.initial;
+  late T _value;
+
+  @override
+  void initState() {
+    super.initState();
+    _value = widget.value ?? widget.initialValue;
+    SchedulerBinding.instance.addPostFrameCallback((_) {
+      setState(() {
+        _value = widget.valueNotifier?.value ?? widget.value ?? _value;
+        _triggered = switch (_triggered) {
+          _ValueUpdaterStatus.off => _ValueUpdaterStatus.on,
+          _ValueUpdaterStatus.on => _ValueUpdaterStatus.off,
+          _ValueUpdaterStatus.initial => _ValueUpdaterStatus.on,
+        };
```

Comment:
Can _triggered ever reach the `off` state? Seems like it starts out as `initial` and then this post frame callback triggers once to move it to `on`, but it appears there is no logic to move it to `off`?